File: test/ado.net/postgresql.cs
// Copyright (c) David Muse
// See the file COPYING for more information.

using System;
using SQLRClient;
using System.Data;

namespace SQLRClientTest
{
    class PostgresqlSQLRAdapterTest : SQLRTest
    {
        public static void Main(String[] args)
        {

            // open connection and command
            SQLRelayConnection sqlrcon =
                new SQLRelayConnection(
                "Data Source=sqlrelay:" +
                "9003:/tmp/postgresql.socket;" +
                "User ID=testuser;" +
                "Password=testpassword;" +
                "Retry Time=0;" +
                "Tries=1;" +
                "Debug=false");
            sqlrcon.Open();

            SQLRelayCommand sqlrcom = (SQLRelayCommand)sqlrcon.CreateCommand();


            // drop existing table
            sqlrcom.CommandText = "drop table testtable";
            ExecuteNonQuery(sqlrcom);


            // error sqlstate
            Console.WriteLine("ERROR SQLSTATE:");
            sqlrcom.CommandText = "create table testtable (col1 int)";
            assertEquals(ExecuteNonQuery(sqlrcom), 0);

            // postgresql reports a duplicate table as 42P07, in every locale
            sqlrcom.CommandText = "create table testtable (col1 int)";
            try
            {
                sqlrcom.ExecuteNonQuery();
                assertTrue(false);
            }
            catch (SQLRelayException ex)
            {
                assertEquals(ex.SQLState, "42P07");
            }

            // and an undefined table as 42P01
            sqlrcom.CommandText = "select * from nonexistenttable";
            try
            {
                sqlrcom.ExecuteReader();
                assertTrue(false);
            }
            catch (SQLRelayException ex)
            {
                assertEquals(ex.SQLState, "42P01");
            }

            sqlrcom.CommandText = "drop table testtable";
            assertEquals(ExecuteNonQuery(sqlrcom), 0);
            Console.WriteLine("\n");


            sqlrcon.Close();

            reportTestStatus();

            Environment.Exit(status);
        }
    }
}

File: test/pythondb/postgresql.py
#! /usr/bin/env python

# Copyright (c) David Muse
# See the file COPYING for more information.

from SQLRelay import PySQLRDB
import sys


def main():

	# instantiation
	print("INSTANTIATION")
	con=PySQLRDB.connect(
		"sqlrelay",9003,
		"/tmp/postgresql.socket",
		"testuser","testpassword",
		0,1,
		# autocommit on, so an expected error doesn't leave
		# postgresql's transaction aborted, with every statement
		# after it failing with 25P02
		autocommit='yes')
	cur=con.cursor()
	print()
	print()


	# error sqlstate
	print("ERROR SQLSTATE")
	try:
		cur.execute("drop table testtable")
	except (PySQLRDB.DatabaseError):
		pass
	cur.execute("create table testtable (col1 int)")

	# postgresql reports a duplicate table as 42P07, in every locale
	try:
		cur.execute("create table testtable (col1 int)")
		raise Exception("the duplicate create table should have failed")
	except (PySQLRDB.DatabaseError) as de:
		print(de.sqlstate)
		assert de.sqlstate=="42P07"

	# and an undefined table as 42P01
	try:
		cur.execute("select * from nonexistenttable")
		raise Exception("the select should have failed")
	except (PySQLRDB.DatabaseError) as de:
		print(de.sqlstate)
		assert de.sqlstate=="42P01"

	cur.execute("drop table testtable")
	print()
	print()


	# error sqlstate - executemany
	print("ERROR SQLSTATE - EXECUTEMANY")
	try:
		cur.executemany("insert into nonexistenttable values ($1)",
					[{'1':1}])
		raise Exception("the insert should have failed")
	except (PySQLRDB.DatabaseError) as de:
		print(de.sqlstate)
		assert de.sqlstate=="42P01"
	print()
	print()


	cur.close()
	con.close()
	del cur
	del con

if __name__ == '__main__':
	main()
	sys.exit(0)
